cardano-crypto-tests/src/Test/Crypto/DSIGN.hs

```diff
@@ -11,17 +11,122 @@ module Test.Crypto.DSIGN
   )
 where
 
+import Text.Show.Pretty (ppShow)
+import Control.Monad (replicateM)
+import Data.ByteString (ByteString)
+import qualified GHC.Exts as GHC
+import qualified Test.QuickCheck.Gen as Gen
+import qualified Crypto.Secp256k1 as SECP
+import Data.Kind (Type)
 import Data.Proxy (Proxy (..))
-
-import Cardano.Crypto.DSIGN
-import Cardano.Crypto.Util (SignableRepresentation(..))
-
-import Test.Crypto.Util hiding (label)
+import Cardano.Crypto.DSIGN (
+  MockDSIGN, 
+  Ed25519DSIGN, 
+  Ed448DSIGN,
+  SECP256k1DSIGN,
+  DSIGNAlgorithm (VerKeyDSIGN,
+                  SignKeyDSIGN,
+                  SigDSIGN,
+                  ContextDSIGN,
+                  Signable,
+                  rawSerialiseVerKeyDSIGN,
+                  rawDeserialiseVerKeyDSIGN,
+                  rawSerialiseSignKeyDSIGN,
+                  rawDeserialiseSignKeyDSIGN,
+                  rawSerialiseSigDSIGN,
+                  rawDeserialiseSigDSIGN),
+  sizeVerKeyDSIGN,
+  sizeSignKeyDSIGN,
+  sizeSigDSIGN,
+  encodeVerKeyDSIGN,
+  decodeVerKeyDSIGN,
+  encodeSignKeyDSIGN,
+  decodeSignKeyDSIGN,
+  encodeSigDSIGN,
+  decodeSigDSIGN,
+  signDSIGN,
+  deriveVerKeyDSIGN,
+  verifyDSIGN,
+  genKeyDSIGN,
+  seedSizeDSIGN,
+  )
+import Cardano.Binary (FromCBOR, ToCBOR)
+import Test.Crypto.Util (
+  Message,
+  prop_raw_serialise,
+  prop_size_serialise,
+  prop_cbor_with,
+  prop_cbor,
+  prop_cbor_size,
+  prop_cbor_direct_vs_class,
+  prop_no_thunks,
+  arbitrarySeedOfSize
+  )
 import Test.Crypto.Instances ()
-import Test.QuickCheck ((=/=), (===), (==>), Arbitrary(..), Gen, Property)
+import Test.QuickCheck (
+  (=/=), 
+  (===), 
+  Arbitrary(..), 
+  Gen, 
+  Property,
+  forAllShow
+  )
 import Test.Tasty (TestTree, testGroup)
 import Test.Tasty.QuickCheck (testProperty)
 
+-- Captures ways of generating each of the associated types of DSIGN
```

Comment:
Why is it necessary to also provide a generalization of genVK and genSK, when we never use it? Can we not simply pass a `Gen (SigDSIGN v)`?

Reply:
Oh, and this design also prevents shrinking, doesn't it?

Reply:
We do use them. Consider line ranges 167-173, 180-187, 195-201, 208-209, 213-214, 218-223 at least. We also have several derived generators using one or both.

As far as preventing shrinking goes - we can't shrink anyway for almost anything, and indeed, the original design _also_ prevented shrinking, since all shrinkers were defined as `const []`. My changes have not removed any functionality or capability that existed before, but _added_ plenty.

Reply:
Ah, sorry about the confusion - yes, we are using the (quasi-)methods, but we are not using the ability to generalize - all the methodologies we define here use the same implementations for `genVerKey` and `genSignKey`, so we might as well just use `default{Ver,Sign}KeyGen`.

Reply:
If we did what I suggest below then we would need `genSignKey`, but perhaps still not `genVerKey`. Not sure it costs us much, though.

Reply:
@tdammers I'm happy to specialize in the way you describe. I don't think it's a particularly bad idea to have the _ability_ to generalize in the way the verification and sign keys are generated, even if we don't use it currently, but I'm not strongly wedded to this.

Reply:
I totally understand the thinking here, it just tickles my YAGNI nerve. I would personally start with only generalizing what needs to be generalized, which, at this point, is only `genSig` bit.

And, upon further thought, what we're *really* generalizing here isn't even generating a *signature*, but generating a *message* to sign. If you compare the `defSigGen` function with the `go` function inside the `secp256k1Methodology`, the only difference is how the message is generated.

So what I would propose is that we ditch the `TypesMethodology` type entirely, and instead just pass a generator for the message type (`genSECPMsg` for secp256k1, `arbitrary` for everything else), and then implement `genSig` in terms of that.